Add obfuscate processor (#4515)
opensearch-project/documentation-website: _data-prepper/pipelines/configuration/processors/obfuscate.md

---
layout: default
title: obfuscate
parent: Processors
grand_parent: Pipelines
nav_order: 71
---

# obfuscate

The `obfuscate` process enables obfuscation to fields inside your documents in order to protect sensitive data. 

## Usage

> In this example, we have a document which contains a `log` field and a `phone` field that contains data we want to obfuscate, as shown in the following object:

Reviewer: ```suggestion
In this example, we have a document that contains a `log` field and a `phone` field. We want to obfuscate the data in both fields, as shown in the following object:
```

> In this example, we have a document that contains a `log` field and a `phone` field. We want to obfuscate the data in both fields, as shown in the following object:

Reviewer 2: For procedures or instructions, ensure that action is taken by the user (“Then you can stop the container...”) rather than the writer (“We also have to stop the container...”).

> In this example, we have a document that contains a `log` field and a `phone` field. We want to obfuscate the data in both fields, as shown in the following object:

Author: ```suggestion
In this example, a document contains a `log` field and a `phone` field, as shown in the following object:
```

```json
{
  "id": 1,
  "phone": "(555) 555 5555",
  "log": "My name is Bob and my email address is abc@example.com"
}
```

> To obfuscate the `log` and `phone` fields multiple fields, add the `obfuscate` processor and call each field in the `source` option, as showing the following example. To account for both the `log` and `phone`, the following examples uses multiple `obfuscate` processors, since each processor can only obfuscate one source.

Reviewer: ```suggestion
To obfuscate the `log` and `phone` fields, add the `obfuscate` processor and call each field in the `source` option. To account for both the `log` and `phone` fields, the following example uses multiple `obfuscate` processors, since each processor can only obfuscate one source.
```

> In the first `obfuscate` processor in the pipeline, the source `log` uses several configuration options to mask the data in the log field. For more details on these options, see [configuration](#configuration).

Reviewer 2: Add ", as shown in the following X" to the first sentence.

> In the first `obfuscate` processor in the pipeline, the source `log` uses several configuration options to mask the data in the log field. For more details on these options, see [configuration](#configuration).

Author: ```suggestion
In the first `obfuscate` processor in the pipeline, the source `log` uses several configuration options to mask the data in the log field, as shown in the following example. For more details on these options, see [configuration](#configuration).
```

```yaml
pipeline:
  source:
    http:
  processor:
    - obfuscate:
        source: "log"
        target: "new_log"
        patterns:
          - "[A-Za-z0-9+_.-]+@([\\w-]+\\.)+[\\w-]{2,4}"
        action:
          mask:
            mask_character: "#"
            mask_character_length: 6
    - obfuscate:
        source: "phone"
  sink:
    - stdout:
```

> When run, the `obfuscate` processor parses the fields into the following output.

Reviewer 2: ```suggestion
When run, the `obfuscate` processor parses the fields into the following output:
```

```json
{
  "id": 1,
  "phone": "***",
  "log": "My name is Bob and my email address is abc@example.com",
  "newLog": "My name is Bob and my email address is ######"
}
```

## Configuration

Use the following configuration options with the `obfuscate` processor.

| Parameter | Required | Description |
| :--- | :---  | :---  |
| `source` | Yes | The source field to obfuscate. |

> | `target` | No | The new field in which to store the obfuscated value and leave the original source field unchanged. When no `target` is provided, the source field updates with the obfuscated value. |

Reviewer: ```suggestion
| `target` | No | The new field in which to store the obfuscated value. This leaves the original source field unchanged. When no `target` is provided, the source field updates with the obfuscated value. |
```

| `patterns` | No | A list of regex patterns that allow you to obfuscate specific parts of a field. Only parts that match the regex pattern will obfuscate. When not provided, the processor obfuscates the whole field. |
| `action` | No | The obfuscation action. As of Data Prepper 2.3, only the `mask` action is supported. |

> You can customize the `mask` action with the following optional configuration options:

Reviewer 2: ```suggestion
You can customize the `mask` action with the following optional configuration options.
```

| Parameter | Default | Description |
| :--- | :---  | :---  |

> `mask_character` | `*` | That character to use when masking. Valid characters are !, #, $, %, &, *, and @. |

Reviewer 2: ```suggestion
`mask_character` | `*` | The character to use when masking. Valid characters are !, #, $, %, &, *, and @. |
```

`mask_character_length` | `3` | The number of characters to mask in the field. The value must be between 1 and 10. |

## Predefined patterns

> When using the `patterns` configuration option, you can use a set of predefined obfuscation patterns for common fields. The `obfuscate` processor supports the following predefined patterns:

Reviewer 2: ```suggestion
When using the `patterns` configuration option, you can use a set of predefined obfuscation patterns for common fields. The `obfuscate` processor supports the following predefined patterns.
```

> You cannot use multiple patterns for one obfuscate processor. Use one obfuscate processor for each pattern.

Reviewer: Use one pattern for each obfuscate processor.
Yes?

> You cannot use multiple patterns for one obfuscate processor. Use one obfuscate processor for each pattern.

Author: ```suggestion
You cannot use multiple patterns for one obfuscate processor. Use one pattern for each obfuscate processor for each pattern.
```

{: .note}

> | Pattern Name          | Examples                                                                                                                                                                      |

Reviewer 2: ```suggestion
| Pattern name          | Examples                                                                                                                                                                      |
```
